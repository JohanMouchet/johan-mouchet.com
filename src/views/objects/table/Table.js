// @flow

import * as React from "react";
import cx from "classnames";
import "./Table.scss";

type Props = {
  align?: "top" | "bottom",
  fixedColumns?: boolean,
  vertical?: boolean,
  variant?: "striped",
  hover?: boolean,
  size?:
    | "sm"
    | "sm-vertical"
    | "sm-horizontal"
    | "lg"
    | "lg-vertical"
    | "lg-horizontal",
  responsive?: boolean,
  attributes?: boolean | number | string,
  children: Node,
};

const Table = ({
  align,
  fixedColumns,
  vertical,
  variant,
  hover,
  size,
  responsive,
  className,
  children,
  ...attributes
}: Props) => {
  const classNames = cx(
    "o-table",
    align && `o-table--align-${align}`,
    fixedColumns && "o-table--fixed-columns",
    vertical && "o-table--vertical",
    variant && `o-table--${variant}`,
    hover && !vertical && "o-table--hover",
    size && `o-table--${size}`,
    responsive && "o-table--responsive-@md",
    className
  );

  return <table className={classNames}>{children}</table>;
};

export default Table;
