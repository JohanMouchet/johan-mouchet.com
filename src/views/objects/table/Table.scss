@import "assets/styles/settings/_settings";
@import "assets/styles/tools/_tools";

/* stylelint-disable selector-max-type, no-descending-specificity */
.o-table {
  $root: &;
  $v-padding: t-vertical-rhythm(0.375) !default;
  $h-padding: $v-padding * 2;
  $scale-factor: 2 !default;

  margin: t-vertical-rhythm(2) 0;
  text-align: left;

  tbody tr:not(:last-child) {
    border-bottom: 1px solid $s-color-gray;
  }

  th,
  td {
    padding: $v-padding $h-padding;
  }

  th {
    font-weight: 500;
    white-space: nowrap;
  }

  thead,
  tfoot {
    color: $s-color-primary;
  }

  thead {
    border-bottom: 2px solid;
  }

  tfoot {
    border-top: 1px solid;
  }

  /** Align **/

  /*** Top ***/
  &--align-top {
    thead,
    tbody,
    tfoot {
      vertical-align: top;
    }
  }

  /*** Bottom ***/
  &--align-bottom {
    thead,
    tbody,
    tfoot {
      vertical-align: bottom;
    }
  }

  /** Equal Columns **/
  &--equal-columns {
    table-layout: fixed;
  }

  /** Vertical **/
  &--vertical {
    text-align: center;

    tbody {
      tr:not(:last-child) {
        border-bottom: 0;
      }

      th:not(:last-child),
      td:not(:last-child) {
        border-right: 1px solid $s-color-gray;
      }
    }
  }

  /** Striped **/
  &--striped {
    tbody {
      tr:not(:last-child) {
        border-bottom-color: transparent;
      }

      tr:nth-child(even) {
        background-color: $s-color-gray-lighter;
      }
    }

    // And vertical
    &#{$root}--vertical {
      tbody {
        /* stylelint-disable max-nesting-depth */
        tr:nth-child(even) {
          background-color: transparent;
        }

        td:nth-child(2n) {
          background-color: $s-color-gray-lighter;
        }

        th:not(:last-child),
        td:not(:last-child) {
          border-right: 0;
        }
        /* stylelint-enable max-nesting-depth */
      }
    }
  }

  /** Hover **/
  &--hover {
    tbody tr:hover {
      background-color: $s-color-gray-lighter;
    }

    // And striped
    &#{$root}--striped {
      tbody tr:hover {
        background-color: $s-color-gray;
      }
    }
  }

  /** Small spacing **/

  /* stylelint-disable max-nesting-depth */
  &--sm {
    thead,
    tfoot {
      th,
      td {
        padding-left: $h-padding / $scale-factor;
        padding-right: $h-padding / $scale-factor;
      }
    }

    tbody {
      th,
      td {
        padding: $v-padding / $scale-factor $h-padding / $scale-factor;
      }
    }

    /*** Vertical ***/
    &-vertical {
      tbody {
        th,
        td {
          padding-top: $v-padding / $scale-factor;
          padding-bottom: $v-padding / $scale-factor;
        }
      }
    }

    /*** Horizontal ***/
    &-horizontal {
      th,
      td {
        padding-left: $h-padding / $scale-factor;
        padding-right: $h-padding / $scale-factor;
      }
    }
  }

  /** Large spacing **/
  &--lg {
    thead,
    tfoot {
      th,
      td {
        padding-left: $h-padding * $scale-factor;
        padding-right: $h-padding * $scale-factor;
      }
    }

    tbody {
      th,
      td {
        padding: $v-padding * $scale-factor $h-padding * $scale-factor;
      }
    }

    /*** Vertical ***/
    &-vertical {
      tbody {
        th,
        td {
          padding-top: $v-padding * $scale-factor;
          padding-bottom: $v-padding * $scale-factor;
        }
      }
    }

    /*** Horizontal ***/
    &-horizontal {
      thead,
      tfoot {
        th,
        td {
          padding-left: $h-padding * $scale-factor;
          padding-right: $h-padding * $scale-factor;
        }
      }

      tbody {
        th,
        td {
          padding-left: $h-padding * $scale-factor;
          padding-right: $h-padding * $scale-factor;
        }
      }
    }
  }
  /* stylelint-enable max-nesting-depth */

  /** Responsive **/
  @media screen and (max-width: map-get($s-breakpoints, md)) {
    &--responsive-\@md {
      display: inline-block;
      max-width: 100%;
      overflow-x: auto;
      white-space: nowrap;
      -webkit-overflow-scrolling: touch;
    }
  }
}
