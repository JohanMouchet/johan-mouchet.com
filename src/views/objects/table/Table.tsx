import clsx, { ClassValue } from "clsx";
import styles from "./Table.module.scss";

export const Table = ({
  align,
  equalColumns,
  vertical,
  variant,
  hover,
  size,
  responsive,
  children,
  className,
  ...props
}: {
  align?: "top" | "bottom";
  equalColumns?: boolean;
  vertical?: boolean;
  variant?: "striped";
  hover?: boolean;
  size?:
    | "sm"
    | "sm-vertical"
    | "sm-horizontal"
    | "lg"
    | "lg-vertical"
    | "lg-horizontal";
  responsive?: boolean;
  children: React.ReactNode;
  className?: ClassValue;
} & Omit<React.HTMLProps<HTMLTableElement>, "size">) => {
  const classNames = clsx(
    styles["o-table"],
    align && styles[`o-table--align-${align}`],
    equalColumns && styles["o-table--equal-columns"],
    vertical && styles["o-table--vertical"],
    variant && styles[`o-table--${variant}`],
    hover && !vertical && styles["o-table--hover"],
    size && styles[`o-table--${size}`],
    responsive && styles["md:o-table--responsive"],
    className
  );

  return (
    <table className={classNames} {...props}>
      {children}
    </table>
  );
};
