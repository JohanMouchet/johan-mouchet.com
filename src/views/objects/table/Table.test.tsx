import { render } from "@testing-library/react";
import { Table } from "./Table";

it("Renders <Table /> without crashing", () => {
  expect(
    render(
      <Table
        align="top"
        equalColumns={false}
        vertical={false}
        variant="striped"
        hover={false}
        size="sm"
        responsive={false}
      >
        <thead>
          <tr>
            <th>Header</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>Cell</td>
          </tr>
        </tbody>
        <tfoot>
          <tr>
            <td>Footer</td>
          </tr>
        </tfoot>
      </Table>
    ).container.firstChild
  ).toMatchSnapshot();
});
