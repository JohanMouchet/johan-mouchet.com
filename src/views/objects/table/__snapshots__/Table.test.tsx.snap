// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders <Table /> without crashing 1`] = `
<table
  className="o-table"
/>
`;
